### DISCUSSION TITLE
[Question] could i ask how to use "set_joint_effort_target" ?

### DISCUSSION THREAD

**Discussion by songwookim** [isaac-sim/IsaacLab#2344] [Q&A] [answered]
### Question

Basic questions, related to robot learning, that are not bugs or feature requests will be closed without reply, because GitHub issues are not an appropriate venue for these.

Advanced/nontrivial questions, especially in areas where documentation is lacking, are very much welcome.

For questions that are related to running and understanding Isaac Sim, please post them at the official [Isaac Sim forums](https://docs.omniverse.nvidia.com/app_isaacsim/app_isaacsim/isaac_sim_forums.html).

```
def run_simulator(sim: sim_utils.SimulationContext, entities: dict[str, Articulation], origins: torch.Tensor):
    """Runs the simulation loop."""
    # Define simulation stepping
    sim_dt = sim.get_physics_dt()
    sim_time = 0.0
    robot : Articulation= entities["ur10_dclaw"]
    robot.actuators["arm"].stiffness = torch.zeros_like(robot.actuators["arm"].stiffness ).cuda() # for effort control
    robot.actuators["arm"].damping = torch.zeros_like(robot.actuators["arm"].damping ).cuda()
    robot.write_joint_stiffness_to_sim(robot.actuators["arm"].stiffness)
    robot.write_joint_damping_to_sim(robot.actuators["arm"].damping)
    robot.actuators["gripper"].stiffness = torch.zeros_like(robot.actuators["gripper"].stiffness ).cuda() # for effort control
    robot.actuators["gripper"].damping = torch.zeros_like(robot.actuators["gripper"].damping ).cuda()
        
    i = 6
    while simulation_app.is_running():
        # reset
        
        if count % 150 == 0:
            # reset counters
            sim_time = 0.0
            count = 0
            # reset the scene entities
            
            root_state = robot.data.default_root_state.clone()
            
            root_state[:, :3] += origins[0]
        
            robot.write_root_link_pose_to_sim(root_state[:, :7])
            robot.write_root_com_velocity_to_sim(root_state[:, 7:])
            
            # set joint positions
            default_joint_pos, default_joint_vel = robot.data.default_joint_pos.clone(), robot.data.default_joint_vel.clone()
            robot.set_joint_position_target(default_joint_pos)
            robot.set_joint_velocity_target(default_joint_vel)
            robot.write_joint_state_to_sim(default_joint_pos, default_joint_vel)
            robot.set_joint_effort_target(torch.zeros_like(robot.data.joint_pos[:,:], device=sim.device))
            
            if i > 14:
                i = 6
                
            joint_ids = [i]
            target = torch.full((1, 1), 1.)  # shape (N_envs, 1)

            robot.set_joint_effort_target(target=target, joint_ids=joint_ids)
            print(f"[INFO]: Resetting robots state... {i}th joint  {robot.joint_names[i]}")
            
            robot.write_data_to_sim()
            robot.reset()
            robot.update(sim_dt)
            
            i += 1
            
            # clear internal buffers
            robot.reset()

        # perform step
        sim.step()
        robot.update(sim_dt)
        # update sim-time
        sim_time += sim_dt
        count += 1

```

https://github.com/user-attachments/assets/cfed7b64-7fea-4149-a592-b5b0eb61a03a

First of all, thank you very much for providing IsaacLab — I truly appreciate your work.

Above, I’ve included the main code related to execution along with its results.

Below is a video showing the expected motion of the finger gripper.

https://github.com/user-attachments/assets/07015ef9-73b0-4055-a0d4-2a6ba4261391

My question concerns an issue where, after setting an effort target, the robot behaves differently than expected. I’m reaching out to ask for guidance on how to properly debug this situation.

To elaborate: I expected that only the specific joint with the effort target applied would move, but I’m observing that other joints are also moving as if they are coupled. I’m unsure how to resolve this coupling issue and would greatly appreciate your help.

I've attached the robot configuration file, joint information, and the corresponding USD file for reference.

```
UR5_DCLAW_CFG = ArticulationCfg(
    spawn=sim_utils.UsdFileCfg(
        usd_path=f"{WORK_DIR}/custom_assets/Robots/ur5e_dclaw_instanceable_ver5.usd",
        rigid_props=sim_utils.RigidBodyPropertiesCfg(
            disable_gravity=False,
            max_depenetration_velocity=5.0,
        ),
        activate_contact_sensors=True,
    ),
    init_state=ArticulationCfg.InitialStateCfg(
        joint_pos={
            "shoulder_pan_joint": 0.0,
            "shoulder_lift_joint": -1.57,
            "elbow_joint": 0,
            "wrist_1_joint": -1.57,
            "wrist_2_joint": -1.57,
            "wrist_3_joint": 1.57,
            "joint_f1_0": 0.0,
            "joint_f2_0": 0.0,
            "joint_f3_0": 0.0,
            "joint_f1_1": 0.0,
            "joint_f2_1": 0.0,
            "joint_f3_1": 0.0,
            "joint_f1_2": 0.0,
            "joint_f2_2": 0.0,
            "joint_f3_2": 0.0,
        },
    ),
    actuators={
        "arm": ImplicitActuatorCfg(
            joint_names_expr=[".*"],
            velocity_limit=0.0,
            effort_limit=0.0,
            stiffness=800.0,
            damping=40.0,
        ),
        "gripper": ImplicitActuatorCfg(
            joint_names_expr=["joint_f.*"],
            velocity_limit=100.0,
            effort_limit=0.1,
            stiffness=800.0,
            damping=40.0,
        ),
    },
)

```
(change .txt to .usd)
[ur5e_dclaw_instanceable_ver5.txt](https://github.com/user-attachments/files/19825651/ur5e_dclaw_instanceable_ver5.txt)

Lastly, here is a description of my runtime environment.
- issac lab version : 2.0.2
- isaac sim version : 4.5.0
- ubuntu : 22.04

**Comment by RandomOakForest** [comment]
Thank you for posting this and for your interest in Isaac Lab. First be sure you review [the documentation found here](https://isaac-sim.github.io/IsaacLab/main/source/api/lab/isaaclab.assets.html#isaaclab.assets.Articulation.set_joint_effort_target) for `set_joint_effort_target` in the `Articulation` class. Have you considered including a penalty in your reward function for the joint's movement you do not want, accordingly? I will move this to our _Discussions_ section for the team to follow up (which will close this issue for now). Please refer to the newly opened discussion in the future. Thanks.

**Comment by StrainFlow** [comment]
I believe this is the source of your trouble: 

    "arm": ImplicitActuatorCfg(
            joint_names_expr=[".*"],
            velocity_limit=0.0,
            effort_limit=0.0,
            stiffness=800.0,
            damping=40.0,
            
The `joint_names` regular expression is probably matching more joints than you intend and essentially turning them all into one actuator.

**Reply by songwookim** [reply]
@StrainFlow Even after modifying the regular expression, the issue remained the same. As far as I know, the regex only affects how joint actuators are managed in Isaac Lab, so it didn’t seem to make a difference 🥲

**Comment by songwookim** [comment, marked as answer]
We resolved the issue by building the robot model from scratch. Just FYI, the same problem occurs even when using modeling files provided by Omniverse (e.g., Robotiq, UR10)
